Aquarium direct map

JiangYizhou opened the pull request:
This is a native OpenGL port of webgl aquarium. The data structure of this version is similar to the webgl version. Currently, it contains opengl backend and can run on Windows, Linux and Mac. Please take a look when you have time, Thanks. @kenrussell @zhenyao @vonture @null77 @Kangz @kainino0x @Richard-Yunchao @gyagp @qjia7 @Jiawei-Shao @shaoboyan

JiangYizhou committed:
Commit aquarium-native

This is a native OpenGL port of webgl aquarium.

JiangYizhou committed:
Modify source and thirdparty folder path
Rename aquarium-native to aquarium-direct-map

null77 commented on aquarium-native/src/ASSERT.h:
```diff
@@ -0,0 +1,32 @@
+        }                                                                                 \
+    }
+#else
+#define ASSERT(expression) expression
```
consider using the code template that ANGLE and Chromium use:

https://chromium.googlesource.com/angle/angle/+/1c6b26a70f935c14d7294f1eaee74c231b17fc5d/src/common/debug.h#265

The template generates zero extra code while ensuring the expression is referenced. Otherwise there will be additional code generated even with static_cast<void>() on MSVC.

JiangYizhou replied:
Thanks to Jamie, actually we don't intent to integrate ANGLE and Chromium in this direct map version. We can use the assert template of ANGLE and Chromium in the optimized version. We decide to upload two native version of Aquarium. One is direct map version and another is optimized version. The direct map version is more similar to the webgl Aquarium and only has OpenGL backend. The latter optimizes the data structure and will extend to many other backends such as dawn and angle.

null77 commented on aquarium-native/src/ASSERT.h:
```diff
@@ -0,0 +1,32 @@
+        }                                                                                 \
+    }
+#else
+#define SWALLOW_ERROR(expression) expression
```
same comment for this.

null77 commented on aquarium-native/src/ASSERT.h:
```diff
@@ -0,0 +1,32 @@
+#pragma once
```
For all your files, consider including a copyright notice at the top of each file. This seems commonplace practice for open source projects.

JiangYizhou replied:
DONE

null77 commented on aquarium-native/src/Buffer.cpp:
```diff
@@ -0,0 +1,51 @@
+    auto bufferFloat  = attribBuffer->getBufferFloat();
+    auto bufferUShort = attribBuffer->getBufferUShort();
+
+    if (attribBuffer->getType() == "Float32Array")
```
is there a particular reason you are storing types as strings instead of IDs?

JiangYizhou replied:
The string types are read from resource files such as Arch.js. In the direct map version, the types or names are stored as strings to make it more align with the webgl version, and in the optimized version they are stored as C++ type. I'll upload the optimized version soon.

null77 commented on aquarium-native/src/Buffer.cpp:
```diff
@@ -0,0 +1,51 @@
+
+#include "ASSERT.h"
+
+Buffer::Buffer(AttribBuffer *attribBuffer, GLenum target)
```
Best practices are to default init all your member variables.

JiangYizhou replied:
DONE

null77 commented on aquarium-native/src/FPSTimer.h:
```diff
@@ -0,0 +1,27 @@
+
+#include <vector>
+
+static const int NUM_FRAMES_TO_AVERAGE = 16;
```
nit: constexpr

JiangYizhou replied:
DONE

null77 commented on aquarium-native/src/Globals.h:
```diff
@@ -0,0 +1,277 @@
+#include "Scene.h"
+
+#if defined(WIN32) || defined(_WIN32) || defined(__WIN32) && !defined(__CYGWIN__)
+const std::string slash = "\\";
```
can these be constexpr char * instead of const std::string?

JiangYizhou replied:
DONE

null77 commented on aquarium-native/src/Globals.h:
```diff
@@ -0,0 +1,277 @@
+class Texture;
+
+static FPSTimer g_fpsTimer;           // object to measure frames per second;
+static std::unordered_map<std::string, Scene *> g_scenes;  // each of the models
```
is it possible instead to index these maps by ID? I'm not sure exactly why you are using string but just a thought/suggestion.

JiangYizhou replied:
Yes, we removed the maps in the optimized version, and use enum arrays instead.  The js code of WebGL version is something like dictionary, the c++ maps are a direct map for the js code.  
The js code For example,

// Draw Scene
    if (g_sceneGroups.base) {
      DrawGroup(g_sceneGroups.base, genericConst, genericPer);
    }

So in the direct map version, we map the 'g_sceneGroups' as  std::map and the 'base' as a std::string for comparison.

null77 commented on aquarium-native/src/Globals.h:
```diff
@@ -0,0 +1,277 @@
+static std::vector<float> fogColor = {1.0f, 1.0f, 1.0f, 1.0f};
+
+//Generic uniforms
+typedef struct GenericConst_
```
nit: you can probably omit the typedef

JiangYizhou replied:
DONE

null77 commented on aquarium-native/src/Globals.h:
```diff
@@ -0,0 +1,277 @@
+}GenericConst;
+static GenericConst sandConst, genericConst, seaweedConst, outsideConst, innerConst;
+
+typedef struct GenericPer_
```
same

JiangYizhou replied:
DONE

null77 commented on aquarium-native/src/Globals.h:
```diff
@@ -0,0 +1,277 @@
+
+static GenericPer sandPer, genericPer, seaweedPer, outsidePer, innerPer, laserPer, skyPer;
+
+typedef struct SkyConst_
```
same

JiangYizhou replied:
DONE

null77 commented on aquarium-native/src/Globals.h:
```diff
@@ -0,0 +1,277 @@
+
+static SkyConst skyConst;
+
+typedef struct FishPer_
```
and so on

JiangYizhou replied:
DONE

null77 commented on aquarium-native/src/Globals.h:
```diff
@@ -0,0 +1,277 @@
+
+static std::unordered_map<std::string, G_sceneInfo> g_sceneInfoByName;
+
+static G_sceneInfo g_sceneInfo[] = { { "SmallFishA", {"fishVertexShader", "fishReflectionFragmentShader"}, true },
```
initing unordered_map in  a header file is not great. Can you move this to a cpp, and please consider using a const char * or enum to index the map to improve compile time speed. See http://crbug.com/845519 for a similar issue.

JiangYizhou replied:
Move to main.cpp. Some of the strings are compared or modified in the code, so I use the strings for convience, for example,

std::string groupName = g_sceneInfoByName[name.GetString()].group;
            if (groupName == "")
            {
                groupName = "base";
            }
...

 for (auto &v : g_skyBoxUrls)
    {
        std::ostringstream url;
        url << path << sourceFolder << slash << resourceFolder << slash << v;

        v = url.str();
    }
....

std::string floats[3] = {"Big", "Medium", "Small"};    // type
std::string &fishName = fishInfo.name;
                if (fishName.find(type))
                {
                    continue;
                }
We use enum to index the map in optimized version.

JiangYizhou replied:
Resolve this in optimized version

null77 commented on aquarium-native/src/Globals.h:
```diff
@@ -0,0 +1,277 @@
+
+static FishConst fishConst;
+
+static Fish g_fishTable[] = { { "SmallFishA", 1.0f, 1.5f, 30.0f, 25.0f, 10.0f, 0.0f, 16.0f, {10.0f, 1.0f, 2.0f} },
```
similarly for the other tables.

JiangYizhou replied:
Move to main.cpp

null77 reviewed (commented):
Thanks for your work on this. This is great. Some preliminary comments. I only looked at maybe half the files.

JiangYizhou committed:
Refactor code

1.Initialize all class variables by default values
2.Move tables from Global.h to Main.cpp or Scene.cpp
3.Add copyrights on the top of all files
4.Use constexpr char * instead of const std::string
5.Omit the typedef

JiangYizhou committed:
Upload aquarium-optimized

Share resources and shaders among the optimized version and direct map
version.

JiangYizhou committed:
Refactor code of aquarium-optimized

1.Initialize all class variables by default value
2.Add copyrights on the top of all files
3.Use constexpr char * instead of const std::string

JiangYizhou commented:
@kenrussell @zhenyao @vonture @null77 @Kangz @kainino0x @Richard-Yunchao @gyagp @qjia7 @Jiawei-Shao @shaoboyan I've uploaded the optimized version as well. Resources, shaders and third party libs are shared by the two projects. PTAL. Thanks.

kenrussell reviewed (approved):
Thank you @JiangYizhou and others for doing this port and @null77 for reviewing as much as you did. I'm sorry for taking so long to get to it; haven't had time to review everything, but the files I glanced through looked good. Let me merge this now so we can make more progress together.
